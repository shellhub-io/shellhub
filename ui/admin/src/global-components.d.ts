// src/global-components.d.ts
import SnackbarComponent from "./components/Snackbar/Snackbar.vue";

declare module "@vue/runtime-core" {
  export interface GlobalComponents {
    SnackbarComponent: typeof SnackbarComponent;
    RouterLink: typeof import("vue-router")["RouterLink"];
    RouterView: typeof import("vue-router")["RouterView"];
  }
}
