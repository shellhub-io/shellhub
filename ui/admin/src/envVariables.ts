/* eslint-disable import/prefer-default-export */
export const envVariables = {
  announcementsEnable: process.env.SHELLHUB_ANNOUNCEMENTS === "true",
  tinyMceKey: process.env.TINYMCE_API_KEY,
};
