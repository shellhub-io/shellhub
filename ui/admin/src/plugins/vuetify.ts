// Styles
// import "@mdi/font/css/materialdesignicons.css";
import "vuetify/styles";
import { VNumberInput } from "vuetify/labs/VNumberInput";

// Vuetify
import { createVuetify } from "vuetify";

export default createVuetify({
  theme: {
    defaultTheme: "dark",
  },
  components: {
    VNumberInput,
  },
});
