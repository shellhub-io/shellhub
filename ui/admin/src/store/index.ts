import { InjectionKey } from "vue";
import { createStore, Store, useStore as vuexUseStore } from "vuex";
import { stats, StatsState } from "./modules/stats";

import { UsersState, users } from "./modules/users";
import { AuthState, auth } from "./modules/auth";
import { DevicesState, devices } from "./modules/devices";
import { firewallRules, FirewallRulesState } from "./modules/firewall_rules";
import { namespaces, NamespacesState } from "./modules/namespaces";
import { layout, LayoutState } from "./modules/layout";
import { license, LicenseState } from "./modules/license";
import { publicKeys, PublicKeysState } from "./modules/public_keys";
import { sessions, SessionsState } from "./modules/sessions";
import { snackbar, SnackbarState } from "./modules/snackbar";
import { spinner, SpinnerState } from "./modules/spinner";
import { announcement, AnnouncementState } from "./modules/announcement";
import { instance, InstanceState } from "./modules/instance";

export interface State {
  user: UsersState;
  auth: AuthState;
  devices: DevicesState;
  firewallRules: FirewallRulesState;
  layout: LayoutState;
  license: LicenseState;
  namespaces: NamespacesState;
  publicKeys: PublicKeysState;
  sessions: SessionsState;
  snackbar: SnackbarState;
  spinner: SpinnerState;
  stats: StatsState;
  announcement: AnnouncementState;
  instance: InstanceState;
}

// eslint-disable-next-line symbol-description
export const key: InjectionKey<Store<State>> = Symbol();

export const store = createStore<State>({
  modules: {
    users,
    auth,
    devices,
    firewallRules,
    layout,
    license,
    namespaces,
    publicKeys,
    sessions,
    snackbar,
    spinner,
    stats,
    announcement,
    instance,
  },
});

export function useStore(): Store<State> {
  return vuexUseStore(key);
}
