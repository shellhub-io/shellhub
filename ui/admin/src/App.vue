<template>
  <v-app>
    <component :is="layout" :data-test="`${layout}-component`" />
  </v-app>
</template>

<script lang="ts">
import { defineComponent, onMounted, computed } from "vue";
import { useRouter } from "vue-router";
import SimpleLayout from "./layouts/SimpleLayout.vue";
import AppLayout from "./layouts/AppLayout.vue";
import { useStore } from "./store";
import { INotificationsError } from "./interfaces/INotifications";

export default defineComponent({
  name: "App",
  components: {
    appLayout: AppLayout,
    simpleLayout: SimpleLayout,
  },
  setup() {
    const store = useStore();
    const router = useRouter();

    const layout = computed(() => store.getters["layout/getLayout"]);

    const isLoggedIn = computed(() => store.getters["auth/isLoggedIn"]);

    const currentRoute = computed(() => router.currentRoute.value.path);

    onMounted(async () => {
      if (!isLoggedIn.value) {
        try {
          await store.dispatch("auth/logout");
          store.dispatch("layout/setLayout", "simpleLayout");
          router.push("/login");
        } catch {
          store.dispatch("snackbar/showSnackbarErrorAction", INotificationsError.namespaceLoad);
        }
      }

      if (isLoggedIn.value && currentRoute.value !== "/login") {
        const license = await store.dispatch("license/get");
        if (!license || license.expired) {
          store.dispatch("snackbar/showSnackbarErrorAction", INotificationsError.license);
          store.dispatch("layout/setLayout", "appLayout");
          router.push("/license");
        }
        store.dispatch("layout/setLayout", "appLayout");
      }
    });

    return {
      layout,
    };
  },
});
</script>
