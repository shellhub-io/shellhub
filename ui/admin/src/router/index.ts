import { createRouter, createWebHistory } from "vue-router";
import Login from "@admin/views/Login.vue";
import Dashboard from "@admin/views/Dashboard.vue";
import Users from "@admin/views/Users.vue";
import SettingsLicense from "@admin/components/Settings/SettingsLicense.vue";
import SettingsAuthentication from "@admin/components/Settings/SettingsAuthentication.vue";
import Namespaces from "@admin/views/Namespaces.vue";
import Settings from "@admin/views/Settings.vue";

import { store } from "../store";

const routes = [
  {
    path: "/login",
    name: "login",
    component: Login,
  },
  {
    path: "/",
    name: "dashboard",
    component: Dashboard,
  },
  {
    path: "/users",
    name: "users",
    component: Users,
  },
  {
    path: "/user/:id",
    name: "userDetails",
    component: () => import("@admin/views/UserDetails.vue"),
  },
  {
    path: "/devices",
    name: "devices",
    component: () => import("@admin/views/Device.vue"),
    redirect: {
      name: "listDevices",
    },
    children: [
      {
        path: "",
        name: "listDevices",
        component: () => import("@admin/views/Device.vue"),
      },
    ],
  },
  {
    path: "/device/:id",
    name: "deviceDetails",
    component: () => import("@admin/views/DeviceDetails.vue"),
  },
  {
    path: "/sessions",
    name: "sessions",
    component: () => import("@admin/views/Sessions.vue"),
  },
  {
    path: "/session/:id",
    name: "sessionDetails",
    component: () => import("@admin/views/SessionDetails.vue"),
  },
  {
    path: "/settings",
    name: "Settings",
    component: Settings,
    redirect: { name: "SettingProfile" },
    children: [
      {
        path: "authentication",
        name: "SettingAuthentication",
        component: SettingsAuthentication,
      },
      {
        path: "license",
        name: "SettingLicense",
        component: SettingsLicense,
      },
    ],
  },
  {
    path: "/firewall-rules",
    name: "firewall-rules",
    component: () => import("@admin/views/FirewallRules.vue"),
  },
  {
    path: "/firewall-rules/:id",
    name: "firewallRulesDetails",
    component: () => import("@admin/views/FirewallRulesDetails.vue"),
  },
  {
    path: "/namespaces",
    name: "namespaces",
    component: Namespaces,
  },
  {
    path: "/namespace/:id",
    name: "namespaceDetails",
    component: () => import("@admin/views/NamespaceDetails.vue"),
  },
  {
    path: "/announcements",
    name: "announcements",
    component: () => import("@admin/views/Announcements.vue"),
  },
  {
    path: "/announcement/:uuid",
    name: "announcementDetails",
    component: () => import("@admin/views/AnnouncementDetails.vue"),
  },
  {
    path: "/new-announcement",
    name: "new-announcement",
    component: () => import("@admin/views/NewAnnouncement.vue"),
  },
];

const router = createRouter({
  history: createWebHistory("/admin/"),
  routes,
});

router.beforeEach((to, from, next) => {
  if (to.path !== "/login") {
    if (store.getters["auth/isLoggedIn"]) {
      return next();
    }
    return next(`/login?redirect=${to.path}`);
  }
  if (store.getters["auth/isLoggedIn"]) {
    if (to.path === "/login" && to.query.token) {
      return next();
    }
    return next("/");
  }

  return next();
});

export default router;
