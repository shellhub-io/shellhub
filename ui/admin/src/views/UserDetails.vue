<template>
  <div class="d-flex pa-0 align-center">
    <h1>User Details</h1>

    <v-col class="pr-4 text-right">
      <v-tooltip bottom anchor="bottom">
        <template v-slot:activator="{ props }">
          <v-icon tag="a" dark v-bind="props" @click="loginToken"> mdi-login </v-icon>
        </template>
        <span>Login</span>
      </v-tooltip>

      <UserDelete :id="currentUser.id" redirect />
    </v-col>
  </div>

  <v-card v-if="!currentUserIsEmpty" class="mt-2 pa-4">
    <v-card-text>
      <div class="text-overline mt-3">
        <h3>Status:</h3>
        <v-chip
          v-if="currentUser.confirmed === true"
          class="ma-2"
          color="success"
          variant="text"
          prepend-icon="mdi-checkbox-marked-circle"
        >
          Confirmed
        </v-chip>
        <v-chip
          v-else
          class="ma-2"
          color="warning"
          variant="text"
          prepend-icon="mdi-alert-circle"
        >
          Not confirmed
        </v-chip>
      </div>
      <div>
        <div class="text-overline mt-3">
          <h3>uid:</h3>
        </div>
        <div :data-test="currentUser.id">
          <p>{{ currentUser.id }}</p>
        </div>
      </div>

      <div>
        <div class="text-overline mt-3">
          <h3>Email:</h3>
        </div>
        <div :data-test="currentUser.email">
          <p>{{ currentUser.email }}</p>
        </div>
      </div>

      <div>
        <div class="text-overline mt-3">
          <h3>Username:</h3>
        </div>
        <div :data-test="currentUser.username">
          <p>{{ currentUser.username }}</p>
        </div>
      </div>

      <div>
        <div class="text-overline mt-3">
          <h3>Namespaces:</h3>
        </div>
        <div :data-test="currentUser.namespaces">
          <p>{{ currentUser.namespaces }}</p>
        </div>
      </div>
    </v-card-text>
  </v-card>
  <v-card class="mt-2 pa-4" v-else>
    <p class="text-center">Something is wrong, try again !</p>
  </v-card>
</template>

<script lang="ts">
import { computed, defineComponent, onMounted } from "vue";
import { useRoute } from "vue-router";
import { useStore } from "../store";
import UserDelete from "../components/User/UserDelete.vue";
import { INotificationsError } from "../interfaces/INotifications";

export default defineComponent({
  setup() {
    const store = useStore();
    const route = useRoute();

    const userId = computed(() => route.params.id as string);

    onMounted(async () => {
      try {
        await store.dispatch("users/get", userId.value);
      } catch {
        store.dispatch("snackbar/showSnackbarErrorAction", INotificationsError.userDetails);
      }
    });

    const currentUser = computed(() => store.getters["users/user"]);
    const currentUserIsEmpty = computed(() => store.getters["users/user"] && store.getters["users/user"].lenght === 0);

    const loginToken = async () => {
      try {
        const token = await store.dispatch("auth/loginToken", currentUser.value);

        const url = `/login?token=${token}`;
        window.open(url, "_target");
      } catch {
        store.dispatch("snackbar/showSnackbarErrorAction", INotificationsError.errorLoginToken);
      }
    };

    return {
      userId,
      currentUser,
      headers: [
        {
          text: "Id",
          value: "id",
        },
        {
          text: "Email",
          value: "email",
        },
        {
          text: "Username",
          value: "username",
        },
        {
          text: "Namespaces",
          value: "ownedNs",
          align: "center",
        },
      ],
      loginToken,
      currentUserIsEmpty,
    };
  },
  components: { UserDelete },
});
</script>
