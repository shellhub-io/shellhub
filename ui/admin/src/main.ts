import { createApp } from "vue";
import vuetify from "@ui/plugins/vuetify";
import App from "./App.vue";
import { loadFonts } from "./plugins/webfontloader";
import { key, store } from "./store";
import router from "./router";
import SnackbarComponent from "./components/Snackbar/Snackbar.vue";

const app = createApp(App);

loadFonts();

app.use(vuetify);
app.use(router);
app.use(store, key);
app.component("SnackbarComponent", SnackbarComponent);
app.mount("#app");
