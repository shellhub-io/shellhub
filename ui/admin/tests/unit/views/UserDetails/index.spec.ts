import { createStore } from "vuex";
import { createVuetify } from "vuetify";
import { mount, VueWrapper } from "@vue/test-utils";
import { beforeEach, describe, expect, it, vi } from "vitest";
import { key } from "../../../../src/store";
import routes from "../../../../src/router";
import UserDetails from "../../../../src/views/UserDetails.vue";

const user = {
  confirmed: true,
  created_at: "2022-04-13T11:42:49.578Z",
  email: "antony@gmail.com",
  id: "6256b739302b50b6cc5eafcc",
  last_login: "0001-01-01T00:00:00Z",
  name: "antony",
  namespaces: 0,
  password: "15e2b0d3c33891ebb0f1ef609ec419420c20e320ce94c65fbc8c3312448eb225",
  username: "antony",
};

const mockRoute = {
  params: {
    id: "6256b739302b50b6cc5eafcc",
  },
};

describe("User Details", () => {
  const store = createStore({
    state: {},
    getters: {
      "users/user": () => user,
    },
    actions: {
      "users/get": vi.fn(),
      "auth/loginToken": vi.fn(),
      "snackbar/showSnackbarErrorAction": vi.fn(),
    },
  });
  let wrapper: VueWrapper<any>;

  beforeEach(() => {
    const vuetify = createVuetify();

    wrapper = mount(UserDetails, {
      global: {
        plugins: [[store, key], vuetify, routes],
        mocks: {
          $route: mockRoute,
        },
      },
    });
  });

  it("Is a Vue instance", () => {
    expect(wrapper).toBeTruthy();
  });

  it("Renders the component", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Has the correct data", async () => {
    expect(wrapper.vm.currentUser).toEqual(user);
  });

  it("Render the correct title", () => {
    expect(wrapper.find("h1").text()).toEqual("User Details");
  });

  it("Should render the props of the user in the Screen", () => {
    expect(wrapper.find("[data-test='antony']").text()).toContain(user.name);
    expect(wrapper.find("[data-test='antony@gmail.com']").text()).toContain(user.email);
    expect(wrapper.find("[data-test='antony']").text()).toContain(user.username);
    expect(wrapper.find("[data-test='0']").text()).toContain(user.namespaces);
  });

  it("Should render the correct buttons", () => {
    expect(wrapper.find("a.v-icon.mdi-login").exists()).toBe(true);
    expect(wrapper.find("a.v-icon.mdi-delete").exists()).toBe(true);
  });
});
