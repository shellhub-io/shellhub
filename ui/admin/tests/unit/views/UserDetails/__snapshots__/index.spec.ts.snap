// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`User Details > Renders the component 1`] = `
"<div class="d-flex pa-0 align-center">
  <h1>User Details</h1>
  <div class="v-col pr-4 text-right"><a class="mdi-login mdi v-icon notranslate v-theme--light v-icon--size-default v-icon--clickable" role="button" aria-hidden="false" tabindex="0" dark="" aria-describedby="v-tooltip-0"></a>
    <!--teleport start-->
    <!--teleport end--><a class="mdi-delete mdi v-icon notranslate v-theme--light v-icon--size-default v-icon--clickable" role="button" aria-hidden="false" tabindex="0" dark="" aria-describedby="v-tooltip-1"></a>
    <!--teleport start-->
    <!--teleport end-->
    <!---->
    <!---->
  </div>
</div>
<div class="v-card v-theme--light v-card--density-default v-card--variant-elevated mt-2 pa-4">
  <!---->
  <div class="v-card__loader">
    <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
      <!---->
      <div class="v-progress-linear__background"></div>
      <div class="v-progress-linear__buffer" style="width: 0%;"></div>
      <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
        <div class="v-progress-linear__indeterminate">
          <div class="v-progress-linear__indeterminate long"></div>
          <div class="v-progress-linear__indeterminate short"></div>
        </div>
      </transition-stub>
      <!---->
    </div>
  </div>
  <!---->
  <!---->
  <div class="v-card-text">
    <div class="text-overline mt-3">
      <h3>Status:</h3><span class="v-chip v-theme--light text-success v-chip--density-default v-chip--size-default v-chip--variant-text ma-2" draggable="false"><!----><span class="v-chip__underlay"></span>
      <!---->
      <div class="v-chip__prepend"><i class="mdi-checkbox-marked-circle mdi v-icon notranslate v-theme--light v-icon--size-default v-icon--start" aria-hidden="true"></i>
        <!---->
      </div>
      <div class="v-chip__content" data-no-activator=""> Confirmed </div>
      <!---->
      <!----></span>
    </div>
    <div>
      <div class="text-overline mt-3">
        <h3>uid:</h3>
      </div>
      <div data-test="6256b739302b50b6cc5eafcc">
        <p>6256b739302b50b6cc5eafcc</p>
      </div>
    </div>
    <div>
      <div class="text-overline mt-3">
        <h3>Email:</h3>
      </div>
      <div data-test="antony@gmail.com">
        <p>antony@gmail.com</p>
      </div>
    </div>
    <div>
      <div class="text-overline mt-3">
        <h3>Username:</h3>
      </div>
      <div data-test="antony">
        <p>antony</p>
      </div>
    </div>
    <div>
      <div class="text-overline mt-3">
        <h3>Namespaces:</h3>
      </div>
      <div data-test="0">
        <p>0</p>
      </div>
    </div>
  </div>
  <!---->
  <!----><span class="v-card__underlay"></span>
</div>"
`;
